### PR TITLE
8488- Updates for new NATS messages and Resubmit payment action

### PR DESCRIPTION
**Pull request by andrepestana-aot**
*Issue:  bcgov/entity#8488*

*Description of changes:*

- Added config; 
- Initialized the queue;
- Updated the stan client_id everytime nats reconnects; 
- Initalized QueueService with a new event loop; 
- Added nest_asyncio to tests for the queue services work properly.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namex license (Apache 2.0).

### CODE REVIEW

**Comment by severinbeauvais** [comment]
Could you please update the ticket number in the description to the "link format"? This automatically creates a link between this PR and the ticket. Thanks!

File: services/namex-pay/tests/unit/test_worker.py [hunk 1]
```diff
@@ -62,7 +62,7 @@ def test_extract_payment_token():
          'is not',  # end_payment_has_value
          None  # error
          ),
-        ('already draft', Payment.PaymentActions.CREATE.value, State.DRAFT, State.DRAFT, 'N', 'N', datetime.utcnow(), 0, 'COMPLETED', 'COMPLETED', None, 'is', None),
+        ('already draft', Payment.PaymentActions.CREATE.value, State.DRAFT, State.DRAFT, 'N', 'N', datetime.utcnow(), 0, 'COMPLETED', 'COMPLETED', None, 'is not', None),
```
Comment (andrepestana-aot) [inline services/namex-pay/tests/unit/test_worker.py:65]: There was a change in the business logic here.

File: services/namex-pay/src/namex_pay/worker.py [hunk 2]
```diff
@@ -92,7 +93,7 @@ async def update_payment_record(payment: Payment) -> Optional[Payment]:
 
     payment_action = payment.payment_action
     nr = RequestDAO.find_by_id(payment.nrId)
-    if payment_action == Payment.PaymentActions.CREATE.value:  # pylint: disable=R1705
+    if payment_action in [Payment.PaymentActions.CREATE.value, Payment.PaymentActions.RESUBMIT.value]:  # pylint: disable=R1705
```
Comment (severinbeauvais) [inline services/namex-pay/src/namex_pay/worker.py:96]: I know what's going because I'm reviewing this specific PR, but how about a comment here to explain to a future developer what's going on here?

**Review by severinbeauvais — APPROVED** [review]
LGTM but please wait for approvals from developers more familiar with NATS messaging.

**Review by kialj876 — APPROVED** [review]
Nice work, lgtm but I am also unfamiliar with the NATS setup now so this should wait for Thor I think

**Review by vysakh-menon-aot — APPROVED** [review]

**Comment by andrepestana-aot** [comment]
@kialj876 as I added some env vars I think there's something to be configured to openshift as well?

File: api/namex/services/queue.py [hunk 3]
```diff
@@ -107,8 +102,15 @@ async def connect(self):
             if not ctx.nats.is_connected:
                 self.stan_options = {**self.stan_options, **{'nats': ctx.nats}}
                 await ctx.nats.connect(**self.nats_options)
+                # Update the client_id if nats is reconnecting
+                self.stan_options['client_id'] = self.get_client_id()
```
Comment (thorwolpert) [inline api/namex/services/queue.py:106]: Why is this being changed?

File: services/namex-pay/config.py [hunk 4]
```diff
@@ -150,6 +158,14 @@ class TestConfig(Config):
     # We can't run NRO locally for running our tests
     DISABLE_NAMEREQUEST_NRO_UPDATES = int(os.getenv('DISABLE_NAMEREQUEST_NRO_UPDATES', 1))
     DISABLE_NAMEREQUEST_SOLR_UPDATES = int(os.getenv('DISABLE_NAMEREQUEST_SOLR_UPDATES', 0))
+    DISABLE_NAMEREQUEST_NATS_UPDATES = int(os.getenv('DISABLE_NAMEREQUEST_NATS_UPDATES', 1))
+
+    NATS_SERVERS = os.getenv('NATS_SERVERS', 'nats://localhost:4222')
```
Comment (thorwolpert) [inline services/namex-pay/config.py:163]: These are already here from ```Config``` use different values if you want it different in test.

File: services/namex-pay/src/namex_pay/worker.py [hunk 5]
```diff
@@ -25,6 +25,8 @@
 the model to a standalone SQLAlchemy usage with an async engine would need
 to be pursued.
 """
+import asyncio
+import nest_asyncio
```
Comment (thorwolpert) [inline services/namex-pay/src/namex_pay/worker.py:29]: What's the new library for?

File: services/namex-pay/src/namex_pay/worker.py [hunk 6]
```diff
@@ -266,7 +270,8 @@ async def process_payment(pay_msg: dict, flask_app: Flask):
 FLASK_APP = Flask(__name__)
 FLASK_APP.config.from_object(APP_CONFIG)
 db.init_app(FLASK_APP)
-
+queue.init_app(FLASK_APP, asyncio.new_event_loop())
```
Comment (thorwolpert) [inline services/namex-pay/src/namex_pay/worker.py:273]: why is this added?
the worker is already called from a loop in run.py

**Review by thorwolpert — COMMENTED** [review]
i don't think adding the api q to the queue listener is correct.

File: api/namex/services/queue.py (hunk 3 above)
Comment (andrepestana-aot) [reply api/namex/services/queue.py:106]: In namex-api it doesn't reconnect frequently but in namex-pay it reconnects often. In that part of the code it checks if nats is not connected and uses the same stan's client_id. When it happens it raises an error saying that the client_id is already connected.

File: services/namex-pay/config.py (hunk 4 above)
Comment (andrepestana-aot) [reply services/namex-pay/config.py:163]: We need these keys for the QueueService from namex-api for all configs:
    NATS_SERVERS
    NATS_CLIENT_NAME
    NATS_CLUSTER_ID
    NATS_QUEUE
    NATS_NR_STATE_SUBJECT 
    NATS_EMAILER_SUBJECT

File: services/namex-pay/src/namex_pay/worker.py (hunk 5 above)
Comment (andrepestana-aot) [reply services/namex-pay/src/namex_pay/worker.py:29]: Quoting https://github.com/erdewit/nest_asyncio: 
`By design asyncio does not allow its event loop to be nested. This presents a practical problem: When in an environment where the event loop is already running it's impossible to run tasks and wait for the result. Trying to do so will give the error "RuntimeError: This event loop is already running".`

That's the only way I could make it work for namex-pay.

File: services/namex-pay/config.py (hunk 4 above)
Comment (thorwolpert) [reply services/namex-pay/config.py:163]: right, and TestConfig already has those from Config

File: services/namex-pay/src/namex_pay/worker.py (hunk 5 above)
Comment (thorwolpert) [reply services/namex-pay/src/namex_pay/worker.py:29]: but why have you added the namex-api queue to the this listener, and in it's own loop.
solving a problem that maybe shouldn't have been created?

File: services/namex-pay/src/namex_pay/worker.py (hunk 6 above)
Comment (andrepestana-aot) [reply services/namex-pay/src/namex_pay/worker.py:273]: In run.py it runs QueueServiceManager. QueueService is a new queue manager added to namex-api and needs to be initialized in some way. I don't know if it's possible to change QueueService to work the same way as QueueServiceManager. What do you think?

**Comment by andrepestana-aot** [comment]
> i don't think adding the api q to the queue listener is correct.

What do you mean? Having namex-api as dependency for namex-pay? What do you suggest?

File: services/namex-pay/src/namex_pay/worker.py (hunk 5 above)
Comment (andrepestana-aot) [reply services/namex-pay/src/namex_pay/worker.py:29]: Well, there's a problem. Maybe there's a better way to solve it:

Without adding the queue to namex-pay we got:
```
    raise err
  File "/home/andre/repos/namex/services/namex-pay/venv/lib/python3.8/site-packages/namex/services/queue.py", line 131, in publish_json_to_subject_sync
    self.loop.run_until_complete(self.async_publish_json(payload, subject))
AttributeError: 'NoneType' object has no attribute 'run_until_complete'

```

With:
`queue.init_app(FLASK_APP)`

```
  File "/home/andre/repos/namex/services/namex-pay/venv/lib/python3.8/site-packages/namex/services/queue.py", line 131, in publish_json_to_subject_sync
    self.loop.run_until_complete(self.async_publish_json(payload, subject))
  File "/usr/lib/python3.8/asyncio/base_events.py", line 592, in run_until_complete
    self._check_running()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 552, in _check_running
    raise RuntimeError('This event loop is already running')
RuntimeError: This event loop is already running
```

What do you suggest for this case?

File: services/namex-pay/config.py (hunk 4 above)
Comment (andrepestana-aot) [reply services/namex-pay/config.py]: ok. Got it and removed.

File: api/namex/services/queue.py (hunk 3 above)
Comment (thorwolpert) [reply api/namex/services/queue.py:106]: you've seen that in the logs?
I haven't seen it reconnecting unless the network has an error.
Over the past 14 days it has had 1 reconnect in Prod

**Review by thorwolpert — CHANGES_REQUESTED** [review]
Just a couple of changes.
The other bits we'll refactor out with the upgrades.
